src/Todo/Model.hs

```diff
@@ -7,21 +9,40 @@ module Todo.Model
     , Deadline (..)
     , Tags
     , Todos
+    , defTodo
+    , getByState
+    , getTopPrio
+    , getByTag
+    , tagStatistics
     ) where
 
-import           Data.List.NonEmpty (NonEmpty (..))
-import qualified Data.Ratio         as R
-import qualified Numeric.Natural    as N
+import Data.List (group, sort, sortBy)
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Ord (comparing)
+import Data.Ratio (Ratio)
+import Numeric.Natural (Natural)
+import GHC.Generics (Generic)
+import Data.Hashable (Hashable (..))
+import Data.HashSet (HashSet)
+import qualified Data.HashSet as HSet
+
+import NESet (NonEmptySet (..))
+import qualified NESet
 
-data State = TODO | DONE | FAIL deriving (Eq, Show)
+data State = TODO | DONE | FAIL deriving (Eq, Show, Generic)
+newtype Title = Title (NonEmpty Char) deriving (Eq, Show, Generic)
+-- the higher the priority value, the more urgent the task
+newtype Priority = Priority (Ratio Natural) deriving (Eq, Show, Ord, Generic)
+newtype Tag = Tag (NonEmpty Char) deriving (Eq, Show, Ord, Generic)
+newtype Deadline = Deadline Int deriving (Eq, Show, Ord, Generic)
+type Tags = NonEmptySet Tag
+type Todos = HashSet Todo
 
-newtype Title = Title (NonEmpty Char) deriving (Eq, Show)
--- the lower the priority value, the more urgent the task
-newtype Priority = Priority (R.Ratio N.Natural) deriving (Eq, Show, Ord)
-newtype Tag = Tag (NonEmpty Char) deriving (Eq, Show)
-newtype Deadline = Deadline Int deriving (Eq, Show, Ord)
-type Tags = NonEmpty Tag
-type Todos = [Todo]
+instance Hashable State
```

Comment:
Doesn't putting `Hashable` under `deriving` without having `Generic` instance for `newtype` work? This should work for generalized newtype deriving...